src/game/rendering_graph_node.c

```diff
@@ -164,285 +164,218 @@ ALIGNED16 struct GraphNodeObject *gCurGraphNodeObject = NULL;
-#endif
-
-        while (currList != NULL) {
-#if SILHOUETTE
-            if (renderPhase == RENDER_PHASE_ZEX_SILHOUETTE) {
-                gSPDisplayList(gDisplayListHead++, dl_silhouette_begin);
-            } else if (renderPhase == RENDER_PHASE_ZEX_NON_SILHOUETTE) {
-                gSPDisplayList(gDisplayListHead++, dl_silhouette_end);
-#ifndef DISABLE_AA
-                // Use normal mode list, no AA
+        // Iterate through the layers on the current render phase.
+        for (currLayer = startLayer; currLayer <= endLayer; currLayer++) {
+            // Set 'currList' to the first DisplayListNode on the current layer.
+            currList = node->listHeads[ucode][currLayer];
+#if defined(DISABLE_AA) || !SILHOUETTE
+            // Set the render mode for the current layer.
+            gDPSetRenderMode(gDisplayListHead++, mode1List->modes[currLayer],
+                                                 mode2List->modes[currLayer]);
+#else
+            if (phaseIndex == RENDER_PHASE_NON_SILHOUETTE) {
+                // To properly cover the silhouette, disable AA.
+                // The silhouette model does not have AA due to the hack used to prevent triangle overlap.
                 gDPSetRenderMode(gDisplayListHead++, (mode1List->modes[currLayer] & ~IM_RD),
-                                                        (mode2List->modes[currLayer] & ~IM_RD));
+                                                     (mode2List->modes[currLayer] & ~IM_RD));
             } else {
-                gDPSetRenderMode(gDisplayListHead++, mode1List->modes[currLayer], mode2List->modes[currLayer]);
-#endif
+                // Set the render mode for the current dl.
+                gDPSetRenderMode(gDisplayListHead++, mode1List->modes[currLayer],
+                                                     mode2List->modes[currLayer]);
             }
 #endif
-            gSPMatrix(gDisplayListHead++, VIRTUAL_TO_PHYSICAL(currList->transform),
-                        G_MTX_MODELVIEW | G_MTX_LOAD | G_MTX_NOPUSH);
-            gSPDisplayList(gDisplayListHead++, currList->displayList);
-            currList = currList->next;
+            // Iterate through all the displaylists on the current layer.
+            while (currList != NULL) {
+                // Add the display list's transformation to the master list.
+                gSPMatrix(gDisplayListHead++, VIRTUAL_TO_PHYSICAL(currList->transform),
+                          (G_MTX_MODELVIEW | G_MTX_LOAD | G_MTX_NOPUSH));
+#if SILHOUETTE
+                if (phaseIndex == RENDER_PHASE_SILHOUETTE) {
+                    // Add the current display list to the master list, with silhouette F3D.
+                    gSPDisplayList(gDisplayListHead++, dl_silhouette_begin);
+                    gSPDisplayList(gDisplayListHead++, currList->displayList);
+                    gSPDisplayList(gDisplayListHead++, dl_silhouette_end);
```

Comment:
do these need to be in this while loop? seems like the beginning and end could be before and after the while loop because `phaseIndex` doesnt change?